src/shims/x86/sse41.rs

```diff
@@ -283,11 +283,18 @@ fn round_first<'tcx, F: rustc_apfloat::Float>(
     assert_eq!(dest_len, left_len);
     assert_eq!(dest_len, right_len);
 
-    let rounding = match this.read_scalar(rounding)?.to_i32()? & !0x80 {
-        0x00 => rustc_apfloat::Round::NearestTiesToEven,
-        0x01 => rustc_apfloat::Round::TowardNegative,
-        0x02 => rustc_apfloat::Round::TowardPositive,
-        0x03 => rustc_apfloat::Round::TowardZero,
+    // The fourth bit of `rounding` only affects the SSE status
+    // register, which cannot be accessed from Miri (or from Rust,
+    // for that matter), so we can ignore it.
+    let rounding = match this.read_scalar(rounding)?.to_i32()? & !0b1000 {
+        // When the third bit is 0, the rounding mode is determined by the
+        // first two bits. When the third bit is 1, the rounding mode is
+        // determined by the SSE status register. We do not support that case,
+        // since modifying it is unsupported in Miri (and Rust).
```

Comment:
Is the SSE status register different from the float status register?

For the float status register, we just assume it is always in NearestTiesToEven mode... but that's the rounding mode for float-to-float operations, right? Is the integer rounding mode even controlled in a status register for regular floats?

So for the SSE status register, I assume we'd just say that it is always assumed to have a particular rounding mode. But I don't know which rounding mode e.g. LLVM will assume here.

Reply:
Yes, it is the same register that specifies rounding mode for arithmetic operations, so I guess we can assume to be round-to-nearest (we already do that for float-to-int conversions).

It looks like LLVM does not const-fold it, even when the rounding mode does not depend on the status register.
 https://godbolt.org/z/zdh3McGcb

Reply:
Wait, don't the normal float-to-int cast do a round-to-zero?

Reply:
SSE has two version: one that rounds to zero (`_mm_cvttss_si32`) and one that uses the rounding mode of the status register (`_mm_cvtss_si32`). We assume round-to-nearest for the later.

https://github.com/rust-lang/miri/blob/c8d4e8356525905dc699e38d9ecf35bbd46af2a2/src/shims/x86/sse.rs#L168-L176

Reply:
> SSE has two version: one that rounds to zero (_mm_cvttss_si32) and one that uses the rounding mode of the status register (_mm_cvtss_si32). We assume round-to-nearest for the later.

Ah okay, let's also do that here then.

But independently of that -- `f32 as i32` casts also always round to zero, so I supposed they also ignore the status register? IOW, that status register really only affects SSE operations, not scalar float operations?

Reply:
Yes, Rust's built-in cast will use the same instruction as `_mm_cvttss_si32`.